### PR TITLE
Detect describe block at given line filter and run all tests in it

### PR DESCRIPTION
**Pull request by arnodirlam**
Until now, ExUnit's line filter (such as `mix test path/to/file:123`) only detected a test case starting at the given line number. Running all tests in a describe block was only possible using `mix test --only describe:"String.capitalize/1"`.

This PR adds the possibility to run all tests in a describe block by addressing the line starting it. If `describe "String.capitalize/1"` is at line 123, all tests in it can now be run via `mix test path/to/file:123`.

This filter is shorter and more intuitive as it makes the line filter smarter, and is in line with the behavior of other test runners, such as rspec, as well as what IDEs such as RubyMine expect.

Implementation-wise, I added an automatic tag `@ex_unit_describe_line` that is added to each test's tags as `:describe_line` and then used for filtering.

I've also added a test for existing functionality of setting the `:describe` tag and prepending it to the name (`:test` tag).

Please let me know if this makes sense for you, and if there's anything I can improve :)

### CODE REVIEW

File: lib/ex_unit/lib/ex_unit/case.ex [hunk 1]
```diff
@@ -376,12 +377,14 @@ defmodule ExUnit.Case do
 
       @ex_unit_describe message
       @ex_unit_used_describes message
+      @ex_unit_describe_line __ENV__.line
```
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/case.ex]: Instead of storing a new attribute, we can store the message and the line on ex_unit_describe. :)

File: lib/ex_unit/lib/ex_unit/case.ex [hunk 2]
```diff
@@ -431,11 +434,12 @@ defmodule ExUnit.Case do
     tag = Module.delete_attribute(mod, :tag)
     async = Module.get_attribute(mod, :ex_unit_async)
 
-    {name, describe, describetag} =
+    {name, describe, describe_line, describetag} =
       if describe = Module.get_attribute(mod, :ex_unit_describe) do
```
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/case.ex]: If you change ex_unit_describe to store `{line, describe}`, this `if` needs to be come a `case`. :)

**Comment by josevalim** [comment]
I love this, thank you. One small change and we should be good to go!

**Comment by arnodirlam** [comment]
Okay cool, will do. Could it be better to use a Map or Keyword list for `@ex_unit_describe` to get more flexibility, or is a tuple much better performance-wise? I'll need to check other places where it's used (like with callbacks and `@ex_unit_setup`) and have it done by the weekend.

**Comment by josevalim** [comment]
@arnodirlam I would use a map or keyword list if it was a public API. I would also use a map if we were storing 4+ elements. But a tuple is fine here, especially because there are only two values.

**Comment by arnodirlam** [comment]
Sounds good. I've changed the `@ex_unit_describe` to a tuple. Please have another look, on the code style as well :)

File: lib/ex_unit/lib/ex_unit/callbacks.ex [hunk 3]
```diff
@@ -162,7 +162,8 @@ defmodule ExUnit.Callbacks do
     quote bind_quoted: [var: escape(var), block: escape(block)] do
       name = :"__ex_unit_setup_#{length(@ex_unit_setup)}"
       defp unquote(name)(unquote(var)), unquote(block)
-      @ex_unit_setup [{name, @ex_unit_describe} | @ex_unit_setup]
```
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/callbacks.ex]: Instead of doing the unmatching here, keep this change and the one below exactly as is and instead change [this line](https://github.com/arnodirlam/elixir/blob/3598d65ff0f1681bb5047672b57023ceda4a5804/lib/ex_unit/lib/ex_unit/callbacks.ex#L413) to be like this:

    defp compile_merge({callback, {_line, describe}}) do

```elixir

File: lib/ex_unit/lib/ex_unit/case.ex [hunk 4]
```diff
@@ -324,9 +324,10 @@ defmodule ExUnit.Case do
         end
       end
 
-  When using Mix, you can run all tests in a describe block as:
+  When using Mix, you can run all tests in a describe block (starting at line 123) as either:
 
       mix test --only describe:"String.capitalize/1"
+      mix test path/to/file:123
```
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/case.ex]: I think we need to clarify this requires the *exact* line. So maybe we should rephrase this to:

```
When using Mix, you can run all tests in a describe block by name:

    mix test --only describe:"String.capitalize/1"

or by passing the exact line the describe block starts:

    mix test path/to/file:123
```

Thoughts?

Comment (arnodirlam) [reply lib/ex_unit/lib/ex_unit/case.ex]: I agree, it's more detailed and clear this way.

File: lib/ex_unit/lib/ex_unit/callbacks.ex (hunk 3 above)
Comment (arnodirlam) [reply lib/ex_unit/lib/ex_unit/callbacks.ex]: Okay great! Thanks for taking the time to guide me to the best way to do this. Is there anything else that could be improved?

**Comment by josevalim** [comment]
I will review more carefully when I am back at the computer but it looks
great!
-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

**Comment by whatyouhide** [comment]
Thanks @arnodirlam! 💟

**Comment by arnodirlam** [comment]
My pleasure! Thanks for the great review 😄
